### PR TITLE
fix(app, react-api-client): handle failed analysis for RTP protocol on ODD

### PR DESCRIPTION
**Pull request by ncdiehl11**
[RQA-2673](https://opentrons.atlassian.net/browse/RQA-2673)

# Overview

RTP protocols introduce a new scenario where an RTP protocol can be sent to a Flex, analyze successfully with default values, and then have parameters input that will cause a failed analysis.

To handle this, the ProtocolSetup on the ODD should promp the user with a failed analysis modal with call to action return to protocol setup to reselect parameters and start a new run. Keeping consistent with our current behavior for protocols that fail analysis, the user will still have the option to dismiss the re-parameterization CTA and proceed with the run.

In addition, we need to modify the ProtocolCard on ODD to search not only for the most recent analysis, but the most recent successful analysis, should one exist.

TODO
- [ ] add unit test for new React api client hook

https://github.com/Opentrons/opentrons/assets/47604184/e1745f50-33f6-429f-ac31-1876d5d56f00

# Test Plan

### A. Following analysis failed modal CTA
1. Push this branch to the ODD with `make -C app-shell-odd push-ot3 host={IP}`
2. Send an RTP protocol that will pass analysis with default values, but will fail analysis with a certain selection of parameters. You can use [this protocol](https://opentrons.atlassian.net/browse/RQA-2673)
3. Move to the ODD and confirm that the protocol was received
4. Click the protocol and proceed to parameter selection
5. Make a selection that will caused a failed analysis (if using the example protocol linked here, set the parameter titled "For fail on analysis" to "To fail"
6. Confirm values and confirm that ProtocolSetup page prompts you with AnalysisFailedModal
7. Verify that the call to action button for "Restart setup" pushes you back to the ProtocolDetails page where you can reselect parameters

### B. Ignoring analysis failed modal CTA
1. Repeat steps A1-6
2. Select 'x' button in top right of modal
3. Confirm that you can still run the protocol despite failed analysis

### C. Protocol card finds successful analysis if one exists
1. Repeat steps A1-6
2. Select back button to view protocols dashboard
3. Confirm that the protocol card does not show failed analysis because even though the most recent analysis failed, the protocol was analyzed successfully initially.

# Changelog

- write new React api client hook for querying the most recent successful analysis of a protocol as document
- implement new useMostRecentSuccessfulAnalysisAsDocumentQuery hook in ProtocolCard to find a successful analysis if one exists
- update AnalysisFailedModal to handle null protocolId
- show AnalysisFailedModal at ProtocolSetup if analysis failed
- update tests

# Review requests

auth js

# Risk assessment

medium

[RQA-2673]: https://opentrons.atlassian.net/browse/RQA-2673?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/src/organisms/ProtocolSetupParameters/AnalysisFailedModal.tsx [hunk 1]
```diff
@@ -36,7 +36,7 @@ export function AnalysisFailedModal({
   }
 
   const handleRestartSetup = (): void => {
-    history.push(`/protocols/${protocolId}`)
+    history.push(protocolId != null ? `/protocols/${protocolId}` : '/protocols')
```
Comment (koji) [inline app/src/organisms/ProtocolSetupParameters/AnalysisFailedModal.tsx:39]: Could you add a test case for `null`?

File: app/src/pages/ProtocolSetup/index.tsx [hunk 2]
```diff
@@ -872,6 +882,15 @@ export function ProtocolSetup(): JSX.Element {
...
     <>
+      {showAnalysisFailedModal &&
+        analysisErrors != null &&
+        analysisErrors?.length > 0 && (
+          <AnalysisFailedModal
+            setShowAnalysisFailedModal={setShowAnalysisFailedModal}
+            protocolId={runRecord?.data.protocolId ?? null}
+            errors={analysisErrors.map(error => error.detail)}
+          />
+        )}
```
Comment (koji) [inline app/src/pages/ProtocolSetup/index.tsx]:
```suggestion
      {showAnalysisFailedModal &&
        analysisErrors != null &&
        analysisErrors?.length > 0  ? (
          <AnalysisFailedModal
            setShowAnalysisFailedModal={setShowAnalysisFailedModal}
            protocolId={runRecord?.data.protocolId ?? null}
            errors={analysisErrors.map(error => error.detail)}
          />
        ) : null}
```

File: react-api-client/src/protocols/useMostRecentSuccessfulAnalysisAsDocumentQuery.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import { getProtocolAnalysisAsDocument } from '@opentrons/api-client'
+import { useQuery } from 'react-query'
+import { useHost } from '../api'
+
+import type { UseQueryResult, UseQueryOptions } from 'react-query'
+import type { HostConfig } from '@opentrons/api-client'
+import {
+  CompletedProtocolAnalysis,
+  ProtocolAnalysisSummary,
+} from '@opentrons/shared-data'
```
Comment (koji) [inline react-api-client/src/protocols/useMostRecentSuccessfulAnalysisAsDocumentQuery.ts]:
```suggestion
import { useQuery } from 'react-query'
import { getProtocolAnalysisAsDocument } from '@opentrons/api-client'
import { useHost } from '../api'

import type { UseQueryResult, UseQueryOptions } from 'react-query'
import type { HostConfig } from '@opentrons/api-client'
import type {
  CompletedProtocolAnalysis,
  ProtocolAnalysisSummary,
} from '@opentrons/shared-data'
```

**Review by koji — APPROVED** [review]
Nice work! 🤖 
The branch worked as expected!
The changes look good to me.

File: react-api-client/src/protocols/useMostRecentSuccessfulAnalysisAsDocumentQuery.ts (hunk 3 above)
Comment (ncdiehl11) [reply react-api-client/src/protocols/useMostRecentSuccessfulAnalysisAsDocumentQuery.ts]: Sorry, forgot to rearrange after the VSCode autoimports 😓

Comment (koji) [reply react-api-client/src/protocols/useMostRecentSuccessfulAnalysisAsDocumentQuery.ts]: no worries, that happens to most prs 🤣

**Pull request merged**